### PR TITLE
with_seed() and with_preserve_seed()

### PR DESCRIPTION
**Pull request by krlmlr**
Taken from hadley/ggplot2#1996

@hadley: PTAL.

### CODE REVIEW

File: R/seed.R [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#'
+#' `with_seed()` runs code with a specific random seed and resets it afterwards.
+#'
+#' @template with
+#' @param seed `[integer(1)]`\cr The random seed to use to evaluate the code.
+#'   Pass `NA` to use the RNG to draw the random seed, pass `NULL` to not change
+#'   the random seed at all (and also to avoid resetting it).
+#' @export
+with_seed <- function(seed, code) {
+  if (is.null(seed)) {
```
Comment (jimhester) [inline R/seed.R]: I do not think we should have a `NULL` case, if you don't want to do anything with the seed why would you call `with_seed()` in the first place? (I realize this API simplifies the logic in the ggplot2 example)

File: R/seed.R [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#' Random seed
+#'
+#' `with_seed()` runs code with a specific random seed and resets it afterwards.
+#'
+#' @template with
+#' @param seed `[integer(1)]`\cr The random seed to use to evaluate the code.
+#'   Pass `NA` to use the RNG to draw the random seed, pass `NULL` to not change
+#'   the random seed at all (and also to avoid resetting it).
+#' @export
+with_seed <- function(seed, code) {
```
Comment (jimhester) [inline R/seed.R]: We should probably also expose the `kind` and `normal.kind` parameters to `set.seed()` as well

File: R/seed.R (hunk 1 above)
Comment (krlmlr) [inline R/seed.R]: Agreed, will get rid of this special case. Do we need the NA case?

Comment (jimhester) [inline R/seed.R]: I think not, maybe we could just include an example showing setting a random seed, e.g.

```r
with_seed(sample.int(.Machine$integer.max, 1L), {
  # ... some code
})
```

**Comment by krlmlr** [comment]
Thanks. I've addressed your feedback.

**Comment by jimhester** [comment]
Thanks!

**Pull request merged**